### PR TITLE
feat: serialize AST to json

### LINKED ISSUE #93: Serialize to parse_query output
In Prometheus 3.0, a new API called `parse_query` is implemented for the new PromLen UI. And this UI will be ported to more Prometheus frontend including Perses https://github.com/perses/perses/pull/2344

image

The diagram is backed by a data structure returned from `parse_query` API. A sample output looks like:

```json
{
  "status": "success",
  "data": {
    "args": [
      {
        "matchers": [
          {
            "name": "__name__",
            "type": "=",
            "value": "prometheus_tsdb_wal_writes_failed_total"
          }
        ],
        "name": "prometheus_tsdb_wal_writes_failed_total",
        "offset": 0,
        "range": 60000,
        "startOrEnd": null,
        "timestamp": null,
        "type": "matrixSelector"
      }
    ],
    "func": {
      "argTypes": [
        "matrix"
      ],
      "name": "rate",
      "returnType": "vector",
      "variadic": 0
    },
    "type": "call"
  }
}

```

We will need to implement the `data` part of json above from library to allow downstream users to implement prometheus compatible `parse_query`.

### PR DESCRIPTION
**Pull request by sunng87**
Fixes #93 

This patch attempts to add AST serialization to match output of prometheus 3.0 parse_query.

### CODE REVIEW

**Comment by sunng87** [comment]
Please let me know if there are promql statements have:

- subquery that has `step` is null
- function that has `variadic` true

File: src/parser/ast.rs [hunk 1]
```diff
@@ -197,12 +200,87 @@ impl BinModifier {
...
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Offset {
     Pos(Duration),
     Neg(Duration),
 }
 
+impl Offset {
+    pub fn value_millis(&self) -> i128 {
```
Comment (evenyag) [inline src/parser/ast.rs:267]: Should we make it public?

Consider renaming it to `as_millis()` as it calls duration's `as_millis()`.

File: src/parser/ast.rs [hunk 2]
```diff
@@ -197,12 +200,87 @@ impl BinModifier {
...
+        match self {
+            Self::Pos(dur) => dur.as_millis() as i128,
+            Self::Neg(dur) => -(dur.as_millis() as i128),
+        }
+    }
+
+    #[cfg(feature = "ser")]
+    pub fn serialize_offset<S>(offset: &Option<Self>, serializer: S) -> Result<S::Ok, S::Error>
```
Comment (evenyag) [inline src/parser/ast.rs:275]: `pub(crate)`?

File: src/util/duration.rs [hunk 3]
```diff
@@ -135,6 +135,27 @@ pub fn display_duration(duration: &Duration) -> String {
     ss
 }
 
+#[cfg(feature = "ser")]
+pub fn serialize_duration<S>(dur: &Duration, serializer: S) -> Result<S::Ok, S::Error>
```
Comment (evenyag) [inline src/util/duration.rs:139]: non-blocking: Should we expose these functions?

**Review by evenyag — APPROVED** [review]
Others LGTM. I wonder if we should make some functions `pub(crate)` only.

**Comment by sunng87** [comment]
Sounds good. We should keep minimal access.

**Comment by waynexia** [comment]
> Please let me know if there are promql statements have:
> 
> * subquery that has `step` is null
> * function that has `variadic` true

AFAIK, subquery is a range query that doesn't accept a null step. And existing functions don't have variadic param list

**Review by waynexia — APPROVED** [review]

**Comment by sunng87** [comment]
@waynexia @evenyag could you please cut a release for this?

**Pull request merged**